src/main/java/com/onaio/steps/helper/AuthDialog.java

```diff
@@ -0,0 +1,236 @@
+                            onAuthListener.onAuthFailed(AuthDialog.this);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private void togglePasswordVisibility() {
+        if(passwordEditText.getInputType() == InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD) {//password is showing
+            passwordEditText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
+            togglePasswordVisibility.setText(R.string.show_password);
+        } else {
+            passwordEditText.setInputType(InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
+            togglePasswordVisibility.setText(R.string.hide_password);
+        }
+    }
+
+    /**
+     * This method generates an MD5 hash for the provided password
+     *
+     * @param password  The plaintext password to be hashed
+     * @return  The MD5 hash for the provided password or NULL if an error occurred
+     */
+    public String hashPassword(String password) {
+        try {
+            // Create MD5 Hash
+            MessageDigest digest = java.security.MessageDigest.getInstance("MD5");
+            digest.update(password.getBytes());
+            byte messageDigest[] = digest.digest();
+
+            // Create Hex String
+            StringBuilder hexString = new StringBuilder();
+            for (byte aMessageDigest : messageDigest) {
+                String h = Integer.toHexString(0xFF & aMessageDigest);
+                while (h.length() < 2)
+                    h = "0" + h;
+                hexString.append(h);
+            }
+            return hexString.toString();
+
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * This method returns the long value of the timestamp for when last the settings screen was
+     * authenticated
+     *
+     * @return  The long value of the timestamp or -1 if unable to get the last auth time
+     */
+    public long getLastAuthTime() {
+        long lastAuthTime = -1;
+        String lastAuthTimeString = KeyValueStoreFactory.instance(activity).getString(SETTINGS_AUTH_TIME);
+        if(lastAuthTimeString != null) {
+            try {
+                lastAuthTime = Long.parseLong(lastAuthTimeString);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        return lastAuthTime;
+    }
+
+    /**
+     * This method updates the last authentication time to now
+     */
+    private void updateLastAuthTime() {
+        long now = Calendar.getInstance().getTimeInMillis();
+        KeyValueStoreFactory.instance(activity).putString(SETTINGS_AUTH_TIME, String.valueOf(now));
+    }
+
+    public boolean needsAuth() {
+        long lastAuthTime = getLastAuthTime();
+        long now = Calendar.getInstance().getTimeInMillis();
+        if(lastAuthTime <= now && (now - 30000) < lastAuthTime) {
```

Comment:
True. That's the number of milliseconds before the authorization expires in the settings screen